Public/Logging/Get-Logger.ps1

```diff
@@ -3,23 +3,47 @@
     .SYNOPSIS
     Returns an instance of the logger object.
 
-    .EXAMPLE
+    .EXAMPLE with full log name
     $Logger = Get-Logger -ShowTime -LogPath 'C:\temp\test.log'
     $Logger.AddErrorRecord("test error")
     $Logger.AddInfoRecord("test info")
     $Logger.AddSuccessRecord("test success")
     $Logger.AddRecord("test record")
+
+    .EXAMPLE with directory name and auto-generated log name
+    $Logger = Get-Logger -ShowTime -LogsDir 'C:\temp'
+    $Logger.AddErrorRecord("test error")
+
+    .EXAMPLE with directory name and logo name defined separately
+    $Logger = Get-Logger -ShowTime -Directory 'C:\temp' -Filename 'test.log'
+    $Logger.AddErrorRecord("test error")
+
+    .EXAMPLE without logfile, only console output
+    $Logger = Get-Logger -ShowTime
+    $Logger.AddErrorRecord("test error")
 #>
 
 function Get-Logger {
-    [CmdletBinding()]
+    [CmdletBinding(DefaultParameterSetName="All")]
     param (
-        [Parameter(Mandatory = $false)]
+        [Parameter(Mandatory = $false, ParameterSetName = 'Logpath')]
         [string] $LogPath,
+        [Parameter(Mandatory = $false, ParameterSetName = 'Complexpath')]
+        [string] $LogsDir,
+        [Parameter(Mandatory = $false, ParameterSetName = 'Complexpath')]
+        [string] $Filename,
         [switch] $ShowTime,
         [string] $TimeFormat = 'yyyy-MM-dd HH:mm:ss'
     )
-    if (-not [string]::IsNullOrWhiteSpace($LogPath)) {
+
+    if ($PSCmdlet.ParameterSetName -eq 'Complexpath') {
+        if (-not $Filename) {
+            $Filename = "$([datetime]::Now.ToString($TimeFormat) -replace('[^.\-\w]', '_'))_ADReporting.log"

```

Comment:
Can we add here that if someone provides $FileName = 'MyReport_.<Date>.log it would automatically convert it? I'm using substitution values like that in PSWinDocumentation and I think it would be a nice addition. And maybe even give it for <Year>, <Month> and so on?

Reply:
You can using the standart datetime format strings:
https://docs.microsoft.com/ru-ru/dotnet/standard/base-types/standard-date-and-time-format-strings

I don’t think we should add custom date parsers.

Reply:
I do not know what to do with the postfix of the log file name, I mean this part: `_ADReporting.log`

Reply:
Right, we can reuse specified format, but you don't know what person wants? You don't know whether he/she wants yearly, monthly or daily logs. Or maybe none at all. How would you use standard datetime format string in PowerShell in this instance? How would you explain it to user? 

By default if no parameters are given it shouldn't log to file, only to screen. Therefore it doesn't really matter what is in there. In PSWinReporting config it can be _PSWinReporting.log which user can change.

Reply:
> By default if no parameters are given it shouldn't log to file, only to screen. Therefore it doesn't really matter what is in there.

I got the impression that you're not confirmed that it doesn't work.  This case is presented in the examples in the header of the script. Just try.

> you don't know what person wants? 

About that. 
The report script has a configuration file, you know. The names of the parameters in it match with the parameters of the logger; this is done intentionally, if the user wants to specify the name that he wants, he can do it himself. It will work exactly as you described it.

Config:
```powershell
$LoggerParameters = @{
    ShowTime   = $true
    LogsDir    = 'C:\temp\logs'
    TimeFormat = 'yyyy-MM-dd HH:mm:ss'
}
```

Logger:
```powershell
    [CmdletBinding(DefaultParameterSetName="All")]
    param (
        [Parameter(Mandatory = $false, ParameterSetName = 'Logpath')]
        [string] $LogPath,
        [Parameter(Mandatory = $false, ParameterSetName = 'Complexpath')]
        [string] $LogsDir,
        [Parameter(Mandatory = $false, ParameterSetName = 'Complexpath')]
        [string] $Filename,
        [switch] $ShowTime,
        [string] $TimeFormat = 'yyyy-MM-dd HH:mm:ss'
)
```

Just add one line:
```powershell
$LoggerParameters = @{
    ShowTime   = $true
    LogsDir    = 'C:\temp\logs'
    TimeFormat = 'yyyy-MM-dd HH:mm:ss'
    Filename = "MonthlyReport_$(Get-Date -UFormat '%M.%Y').log"
}
```

Try the examples from this version of the logger, this version cover all the cases.